common/message_algos/src/filters/dedup_filter.rs

```diff
@@ -0,0 +1,54 @@
+//! Message deduplication
+
+use crate::{filters::*, Envelope};
+use std::collections::VecDeque;
+
+/// Test whether two messages are a duplicates.
+pub trait DedupPolicy: Send {
+    type Message: Send + Clone;
+
+    fn is_duplicate(&self, msg1: &Envelope<Self::Message>, msg2: &Envelope<Self::Message>) -> bool;
+}
+
+/// A message deduper
+pub struct DedupFilter<T: Send + Clone> {
+    dedup_policy: Box<dyn DedupPolicy<Message = T>>,
+    max_history_capacity: usize,
+    history: VecDeque<Envelope<T>>,
+}
+
+impl<T: Send + Clone> DedupFilter<T> {
+    pub fn new(
+        dedup_policy: Box<dyn DedupPolicy<Message = T>>,
+        max_history_capacity: usize,
+    ) -> Self {
+        assert!(max_history_capacity > 0);
+        Self {
+            dedup_policy,
+            max_history_capacity,
+            history: VecDeque::with_capacity(max_history_capacity),
+        }
+    }
+}
+
+impl<T: Clone + Send> MessageFilter for DedupFilter<T> {
+    type Message = T;
+
+    fn filter(&mut self, message: &Envelope<Self::Message>) -> FilterDecision {
+        match self
+            .history
+            .iter()
+            .find(|msg| self.dedup_policy.is_duplicate(msg, message))
```

Comment:
Having a linear search here means we have an algo with a quadratic complexity to de-dup the events.

Reply:
isn't it O(n) considering each Dedup check as O(1). Would be happy to use hashing...